### PR TITLE
Add DDEV Configuration for local development of core

### PR DESCRIPTION
**Pull request by bhanu951**

### CODE REVIEW

File: .gitignore
Comment (joachim-n) [inline .gitignore:1]: Is this necessary in the repo? It's ages since I did an install with this template, but doesn't the process just unzip the archive rather than create a git clone?

File: composer.json
Comment (joachim-n) [inline composer.json:40]: Are these updates that are related to the DDEV stuff? 

If not, they should be a separate MR.

File: CORE-DEVELOPMENT-README.md
Comment (joachim-n) [inline CORE-DEVELOPMENT-README.md:3]: Are steps 1-4 needed for DDEV? Can you not use `composer create-project`? That command will do all that.

File: CORE-DEVELOPMENT-README.md
Comment (joachim-n) [inline CORE-DEVELOPMENT-README.md:8]: Could this be added to a section on using DDEV in the main README please?

File: CORE-DEVELOPMENT-README.md
Comment (joachim-n) [inline CORE-DEVELOPMENT-README.md:9]: There's already something about this in the README, but feel free to expand on it if you feel it's necessary.

File: CORE-DEVELOPMENT-README.md
Comment (joachim-n) [inline CORE-DEVELOPMENT-README.md:11]: Same -- add this to the README.

File: composer.json
Comment (bhanu951) [reply composer.json:40]: Done new MR https://github.com/joachim-n/drupal-core-development-project/pull/17 raised

File: .gitignore
Comment (bhanu951) [reply .gitignore:1]: From my understanding it makes a git clone https://github.com/joachim-n/drupal-core-development-project/blob/main/src/ComposerScripts.php#L32 .

File: CORE-DEVELOPMENT-README.md
Comment (bhanu951) [reply CORE-DEVELOPMENT-README.md:8]: Sure, will update it.

File: CORE-DEVELOPMENT-README.md
Comment (bhanu951) [reply CORE-DEVELOPMENT-README.md:9]: sure

File: CORE-DEVELOPMENT-README.md
Comment (bhanu951) [reply CORE-DEVELOPMENT-README.md:11]: ok

File: CORE-DEVELOPMENT-README.md
Comment (bhanu951) [reply CORE-DEVELOPMENT-README.md:3]: once you clone this repo and execute `ddev composer install` drupal core will be downloaded into `repos/drupal` directory but it is having performance affect and taking atleast 2x time to setup.

File: .gitignore
Comment (joachim-n) [reply .gitignore:1]: Yes, that's the git clone at /repos/drupal. My point is that the overall project isn't a git clone when you do `composer create-project`. So putting thing file there is adding a file that has no effect for people installing this as a project.

File: composer.json
Comment (joachim-n) [reply composer.json:40]: Thanks!

File: .gitignore
Comment (bhanu951) [reply .gitignore]: Done, deleted both `.gitignore` and `.gitkeep`

File: CORE-DEVELOPMENT-README.md
Comment (bhanu951) [reply CORE-DEVELOPMENT-README.md]: Done, Moved to the main README

File: CORE-DEVELOPMENT-README.md
Comment (bhanu951) [reply CORE-DEVELOPMENT-README.md]: Done, Moved to the main README

File: CORE-DEVELOPMENT-README.md
Comment (bhanu951) [reply CORE-DEVELOPMENT-README.md]: Done, Moved to the main README

File: .gitignore
Comment (bhanu951) [reply .gitignore]: so `cd repos` and `git clone --branch=10.1.x https://git.drupalcode.org/project/drupal.git drupal`
are mandatory as well.

File: CORE-DEVELOPMENT-README.md
Comment (bhanu951) [reply CORE-DEVELOPMENT-README.md]: ~once you clone this repo and execute ddev composer install drupal core will be downloaded into repos/drupal directory but it is having performance affect and taking atleast 2x time to setup.~

The above comment is invalid. We would be needing ` cd repos` and `git clone --branch=10.1.x https://git.drupalcode.org/project/drupal.git drupal`  mandatory as well. If not below error is thrown

> The `url `supplied for the path (repos/drupal) repository does not exist

File: .ddev/config.yaml
Comment (rfay) [inline .ddev/config.yaml:1]: I recommend leaving `name:` out of something like this. It just confuses people. I don't really normally recommend checking in a whole config.yaml in a template project like this, but it's OK. If you remove `name:` then it takes the name of the directory it's in.

```suggestion
```

File: .ddev/config.yaml
Comment (rfay) [inline .ddev/config.yaml:260]:
```suggestion
```

These instructional comments are not useful here.

File: .ddev/docker-compose.testing.yml
Comment (rfay) [inline .ddev/docker-compose.testing.yml:19]: Please don't use this image. It works on amd64 machines only, doesn't work on mac M1. Please see (and use) https://github.com/drud/ddev-selenium-standalone-chrome (`ddev get drud/ddev-selenium-standalone-chrome`), which is maintained by @weittzman and has automated tests, and with works on arm64.

**Review by rfay — CHANGES_REQUESTED** [review]
As mentioned, please adjust this to use the maintained, tested, and arm64-friendly https://github.com/drud/ddev-selenium-standalone-chrome; you'll probably want to add `ddev get drud/ddev-selenium-standalone-chrome` to your instructions rather than copying from that, which makes yours unmaintainable.

File: .ddev/docker-compose.testing.yml
Comment (bhanu951) [reply .ddev/docker-compose.testing.yml]: @rfay : Thanks for the review.

I like the idea of it but it seems the image doesn't have support for SQLite DB. I tried to raise a PR [#6](https://github.com/drud/ddev-selenium-standalone-chrome/pull/6) for its support. 

I tried to override `SIMPLETEST_DB` env variable in `config.yml` as below 

```
composer_version: "2"
web_environment:
- SIMPLETEST_DB=sqlite://tmp/sites/simpletest/db.sqlite
nodejs_version: "16"
```

but it has not overrided the value defined in `docker-compose.selenium-chrome.yaml` do you have any suggestion on how to handle it ?

Comment (rfay) [reply .ddev/docker-compose.testing.yml]: As discussed over there, `docker-compose.*.yaml` wins. But you can do it with a docker-compose.override.yaml here, or work a little more over there to use `config.*.yaml`

I'm not sure why you want to force the user to use sqlite in the first place though. Why is that? IMO tests are less valid when using a little-used database type.

File: .ddev/config.yaml
Comment (rfay) [inline .ddev/config.yaml:9]:
```suggestion
```

Removing these will make projects (and this repo) easier to maintain.

File: .ddev/config.yaml
Comment (rfay) [inline .ddev/config.yaml:18]: Removing this will make these easier to maintain. Just accept defaults.
```suggestion
```

**Review by rfay — CHANGES_REQUESTED** [review]
Removing things that should default will make both projects in general and this repo easier to maintain.

**Comment by joachim-n** [comment]
@rfay Thanks so much for reviewing this! I don't know anything at all about DDEV configuration, so I really appreciate your input here!

Happy to merge this when you give it the green light :)

File: .ddev/docker-compose.testing.yml
Comment (bhanu951) [reply .ddev/docker-compose.testing.yml]: > I'm not sure why you want to force the user to use sqlite in the first place though. Why is that? IMO tests are less valid when using a little-used database type.

I am not forcing to use SQLite. I am trying to provide an option for the users to override easily .
And also SQLIte DB tests are faster in compared to MySQL DB, sometimes simpletests leave tables orphaned when the tests failed in MySQL. As we are using our main MySQL DB for tests as well, orphaned tables are present in the main DB itself, which doesn't look good.

File: .ddev/config.yaml
Comment (rfay) [reply .ddev/config.yaml:9]: These are things that people rarely change. And of course, just running `ddev config` will fully populate all this again. What I'm trying to say is, "If you're going to add a .ddev/config.yaml, only add value in it". In general, anybody who uses DDEV knows they can do a `ddev config --auto` and they'll already get all this done for them.

Comment (rfay) [reply .ddev/config.yaml:9]: That said, this is just a matter of personal preference, my usual recommendation. It's not critical.

Comment (pyrello) [reply .ddev/config.yaml:9]: FWIW, I agree with removing these. Were they ever to change in DDEV, this file would need to be updated to new values. If they aren't included, that would just happen in the background. I think it makes this project easier to maintain in the long run. Also, if it turns out this is a big problem with new users, then this can be changed later to add these variables back in.

**Comment by rfay** [comment]
One comment before I test @bhanu951 - When you do a PR please create a new branch for your PR. Using `main` (which also matches the target DB's default branch) will just cause you and the target repo's maintainer pain forever. And it will confuse you when you go to make a new PR.

File: README.md
Comment (rfay) [inline README.md:242]: I don't think this is the right approach at all. Instead, 

```
mkdir drupal-core-development-project && cd drupal-core-development-project
ddev composer create -y --no-install joachim-n/drupal-core-development-project
ddev composer install
ddev get drud/ddev-selenium-standalone-chrome
```

This will make it far more like ordinary DDEV project management, like the quickstarts in [docs](http://127.0.0.1:8000/en/latest/users/quickstart/#drupal)

**Review by rfay — CHANGES_REQUESTED** [review]
Could you please rethink and retest this using `ddev composer create` @bhanu951 ? When you do that, you can completely remove the .ddev/config.yaml, as it will be created by the `ddev config` step.

**Comment by bhanu951** [comment]
> One comment before I test @bhanu951 - When you do a PR please create a new branch for your PR. Using `main` (which also matches the target DB's default branch) will just cause you and the target repo's maintainer pain forever. And it will confuse you when you go to make a new PR.

Sure @rfay : Sure I will keep this in mind. I havent realised I worked on main branch while raising PR. After realising there were already review comments, so left it to preserve review comments.

I raise other PR from a new branch.

**Comment by rfay** [comment]
@bhanu951 you don't need to raise another PR, that will just make a different mess. I was just talking about for the future.

File: README.md
Comment (bhanu951) [reply README.md:242]: Hi @rfay , I tried above instructions and observed below .

1. `ddev composer create -y --no-install joachim-n/drupal-core-development-project` is not creating project it is asking to initialise a project by running `ddev config` which is a additional step.

> could not find a project in /Users/user1/docker/drupal-core-development-project. Have you run 'ddev config'? Please specify a project name or change directories: no .ddev/config.yaml file was found in this directory or any parent

2. It asks for `Docroot Location (current directory):` which user wont be aware of.
3. You can not specify core version to download from repo
4. It is taking atleast 2x time to setup project if you are lucky enough it get finished .
5. At last there is a message stating `Moving install to composer root` which is taking forever to complete.
6. It gives a bad experience to the user.

So IMHO current instructions in README servers better.

Comment (rfay) [reply README.md:242]: You have to create a directory and `cd` into the directory and `ddev config`. This is standard stuff, maybe you can experiment with the [quickstarts](https://ddev.readthedocs.io/en/latest/users/quickstart/#drupal) and see how they work, pattern it on them. Normally for Drupal, `ddev config --docroot=web --project-type=drupal10 --create-docroot`

Current instructions are not normal DDEV stuff and use composer on the host, which may be a different PHP version etc. Please try again.

Comment (bhanu951) [reply README.md:242]: Ok, do let me know if you have any particular steps that you would like to have in particular manner in readme, if you wish to update it I am more than happy. Thanks for all your review comments so far.

**Comment by rfay** [comment]
I hope to get a look at this this week now that DDEV v1.21.4 is out. Thanks for your patience @bhanu951

File: .devcontainer/Dockerfile
Comment (rfay) [inline .devcontainer/Dockerfile:1]: It seems to me like doing a devcontainer here, while admirable, is out of scope. Could this be done in a separate PR?

**Review by rfay — CHANGES_REQUESTED** [review]
Overall, I think the best technique here @bhanu951 is for you to create and maintain a [DDEV add-on](https://ddev.readthedocs.io/en/latest/users/extend/additional-services/) that does all the (lovely) things you want to do. That will give you the ability to maintain the custom commands as time goes on, and will give you automated tests to tell when something is going wrong. 

Then you can do a PR here that explains how to do that, with just a few commands (`mkdir drupal-core-dev-project && cd drupal-core-dev-project && ddev config ... && ddev get <your add-on> && ddev composer create ...`)

I'm happy to help, but it's really easy to create an add-on, and you can work on the add-on locally by just `ddev get /path/to/addon`.

Overall, your long-term ability to maintain this yourself will be a real win, and your ability to get feedback from users. I think this is a great initiative though, and the set of custom commands you've created should really help people.

**Pull request closed without merging**
